source/server/configuration_impl.cc

```diff
@@ -38,11 +38,48 @@ bool FilterChainUtility::buildFilterChain(Network::FilterManager& filter_manager
   return filter_manager.initializeReadFilters();
 }
 
-bool FilterChainUtility::buildFilterChain(
-    Network::ListenerFilterManager& filter_manager,
-    const std::vector<Network::ListenerFilterFactoryCb>& factories) {
-  for (const Network::ListenerFilterFactoryCb& factory : factories) {
-    factory(filter_manager);
+class MissingConfigTcpListenerFilter : public Network::ListenerFilter,
+                                       public Logger::Loggable<Logger::Id::filter> {
+public:
+  MissingConfigTcpListenerFilter() = default;
+
+  // Network::ListenerFilter
+  Network::FilterStatus onAccept(Network::ListenerFilterCallbacks& cb) override {
+    cb_ = &cb;
+    ENVOY_LOG(warn, "Listener filter: new connection accepted while missing configuration. "
+                    "Close socket and stop the iteration onAccept.");
+    cb_->socket().ioHandle().close();
+    return Network::FilterStatus::StopIteration;
+  }
+  Network::FilterStatus onData(Network::ListenerFilterBuffer&) override {
+    // The socket is already closed onAccept. Just return StopIteration here.
+    return Network::FilterStatus::StopIteration;
+  }
+  size_t maxReadBytes() const override { return 1024; }
+
+private:
+  Network::ListenerFilterCallbacks* cb_{};
+};
+
+bool FilterChainUtility::buildFilterChain(Network::ListenerFilterManager& filter_manager,
+                                          const Filter::ListenerFilterFactoriesList& factories) {
+  bool added_missing_config_filter = false;
+  for (const auto& filter_config_provider : factories) {
+    auto config = filter_config_provider->config();
+    if (config.has_value()) {
+      auto config_value = config.value();
+      config_value(filter_manager);
+      continue;
+    }
+    // If a filter config is missing after warming, stop iteration.
```

Comment:
This condition means there is no default config, also no configuration received.   In such case, the ECDS listener filter is empty. Please check the integration test case: TEST_P(ListenerExtensionDiscoveryIntegrationTest, BasicWithoutWarmingFail) .

When this happens, the current code in the PR is closing the connection.  But, I would like to discuss what's the right behavior. The other options considered is to create a MissingConfig filter, and onAccept() just return Continue, i.e, ignore this empty config listener filter all together.

Create a missing filter is also trying to keep the http filter ECDS and listener ECDS behavior consistent in such case.

Reply:
> This condition means there is no default config, also no configuration received. In such case, the ECDS listener filter is empty. Please check the integration test case: TEST_P(ListenerExtensionDiscoveryIntegrationTest, BasicWithoutWarmingFail) .
> 

May I ask why we allow the Listener becomes `active` when there is no initial config received? This question isn't part of this PR, just curious about the reason behind this.

> When this happens, the current code in the PR is closing the connection. But, I would like to discuss what's the right behavior. The other options considered is to create a MissingConfig filter, and onAccept() just return Continue, i.e, ignore this empty config listener filter all together.
> 

If we just interrupt the creation of listener filters and close the connection, it is pretty straightforward, and less code. Also as above said, going through other valid listener filters becomes meaningless in this case.

> Create a missing filter is also trying to keep the http filter ECDS and listener ECDS behavior consistent in such case.

I'm a little doubt, HTTP filter can do the samething here.

Reply:
HTTP ECDS should be holding the listener in a warming state until config is received without the config. I think this also makes sense here.

That doesn't rule out the possibility of a missing config because of validation. The warming is triggered on any received config, not just valid received config. There are two reasons for that:

* Adding a listener for an existing subscription with a wrong received config. The warming targets are defined per subscription, not per provider.
* Dealing with temporary inconsistency between ECDS and LDS servers. E.g. we want LDS to make progress if ECDS is out of consistency that prevents it from producing a valid config.

We could have a separate warming target per each provider. I'd need more thinking about that.

Reply:
Thanks for the comments.

@soulxu for your question "May I ask why we allow the Listener becomes active when there is no initial config received? " :  Actually, when a Listener receives initial config but the config is rejected due to invalid config, the listener can change to "active" state. I modified my original comment. Please check above again.

Regarding your recommendation to add a if check below and close socket and return: https://github.com/envoyproxy/envoy/blob/8f4c0069f3861631038194065daac47cb52b313b/source/server/active_stream_listener_base.h#L87
I think it is risky to be done within this PR. If it needs to be done, it should be done separately.  WDYT?

Reply:
@yanjunxiang-google  thanks for answer my question!

> 
> @soulxu for your question "May I ask why we allow the Listener becomes active when there is no initial config received? " : Actually, when a Listener receives initial config but the config is rejected due to invalid config, the listener can change to "active" state. I modified my original comment. Please check above again.
> 

Why not waiting for config validation before make the listener as active? Or there also has the case, the upcoming update can be invalid config also.

> Regarding your recommendation to add a if check below and close socket and return:
> 
> https://github.com/envoyproxy/envoy/blob/8f4c0069f3861631038194065daac47cb52b313b/source/server/active_stream_listener_base.h#L87
> 
> 
> I think it is risky to do it within this PR. If it needs to be done, it should be done separately. WDYT?

Yes, sure, I think that is ok for me. I can help with this also.

Reply:
Thanks for the comments @soulxu! 

Regarding " Why not waiting for config validation before make the listener as active? Or there also has the case, the upcoming update can be invalid config also."
I am not an expert. IIUC, an invalid config still make a listener active based on below link. Also, I think there is no guarantee the future config updates are valid:
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/init#:~:text=Once%20cluster%20manager%20initialization%20is%20done%2C%20RDS%20and%20LDS%20initialize%20(if%20applicable).%20The%20server%20waits%20for%20a%20bounded%20period%20of%20time%20for%20at%20least%20one%20response%20(or%20failure)%20for%20LDS/RDS%20requests.%20After%20which%2C%20it%20starts%20accepting%20connections.

For having a separate PR to deal with adding a check in below code and close socket/bailout early, please go ahead: https://github.com/envoyproxy/envoy/blob/8f4c0069f3861631038194065daac47cb52b313b/source/server/active_stream_listener_base.h#L87